Algorithm.Framework/Portfolio/BlackLittermanPortfolioConstructionModel.cs

```diff
@@ -0,0 +1,310 @@
+
+            /// process results
+            if (ret > 0)
+            {
+                var weights = symbols.Zip(W, (sym, w) => new { S = sym, W = w }).ToDictionary(r => r.S, r => r.W);
+                algorithm.Debug(" ### [" + string.Join(",", weights.Keys) + "] = [" + string.Join(",", weights.Values) + "]");
+
+                // Create portfolio targets from the specified insights
+                foreach (var insight in insights)
+                {
+                    var weight = (decimal)weights[insight.Symbol];
+                    targets.Add(PortfolioTarget.Percent(algorithm, insight.Symbol, weight));
+                }
+            }
+            return targets;
+        }
+
+        /// <summary>
+        /// Event fired each time the we add/remove securities from the data feed
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance that experienced the change in securities</param>
+        /// <param name="changes">The security additions and removals from the algorithm</param>
+        public override void OnSecuritiesChanged(QCAlgorithmFramework algorithm, SecurityChanges changes)
+        {
+            // clean up data for removed securities
+            foreach (var removed in changes.RemovedSecurities)
+            {
+                _pendingRemoval.Add(removed.Symbol);
+                ReturnsSymbolData data;
+                if (_symbolDataDict.TryGetValue(removed.Symbol, out data))
+                {
+                    _symbolDataDict.Remove(removed.Symbol);
+                    data.RemoveConsolidators(algorithm);
+                }
+            }
+
+            // initialize data for added securities
+            var addedSymbols = new List<Symbol>();
+            foreach (var added in changes.AddedSecurities)
+            {
+                if (!_symbolDataDict.ContainsKey(added.Symbol))
+                {
+                    var symbolData = new ReturnsSymbolData(algorithm, added.Symbol, _lookback, _resolution);
+                    _symbolDataDict[added.Symbol] = symbolData;
+                    addedSymbols.Add(symbolData.Symbol);
+                }
+            }
+
+            if (addedSymbols.Count <= 0)
+                return;
+
+            // warmup our indicators by pushing history through the consolidators
+            algorithm.History(addedSymbols, _lookback, _resolution)
+                .PushThrough(bar =>
+                {
+                    ReturnsSymbolData symbolData;
+                    if (_symbolDataDict.TryGetValue(bar.Symbol, out symbolData))
+                    {
+                        symbolData.ROC.Update(bar.EndTime, bar.Value);
+                        if (symbolData.IsReady)
+                        {
+                            var values = symbolData.Window.Select(x => x.Value).ToArray();
+                            algorithm.Log(" ### " + symbolData.Symbol.Value + string.Join(",", values));
```

Comment:
In the previous peer-review, I forgot to ask to change this `Log` call for `Debug` call.